terraform/provider/provider.go

```diff
@@ -37,42 +42,100 @@ func Provider() *schema.Provider {
 			},
 		},
 		ResourcesMap: map[string]*schema.Resource{
-			"teleport_user": resourceTeleportUser(),
 			"teleport_role": resourceTeleportRole(),
+			"teleport_user": resourceTeleportUser(),
 		},
-		ConfigureFunc: providerConfigure,
+		ConfigureContextFunc: configureContext,
 	}
 }
 
-// providerConfigure accepts provider settings schema,
+// configureContext accepts provider settings schema,
 // initializes a Teleport's `auth.Client` and returns it.
-func providerConfigure(d *schema.ResourceData) (interface{}, error) {
-	certPath := d.Get("cert_path").(string)
-	keyPath := d.Get("key_path").(string)
-	rootCAsPath := d.Get("root_ca_path").(string)
+func configureContext(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
+	certPath, err := getPath(d, "cert_path")
+	if err != nil {
+		return nil, diag.FromErr(trace.Wrap(err))
+	}
+
+	keyPath, err := getPath(d, "key_path")
+	if err != nil {
+		return nil, diag.FromErr(trace.Wrap(err))
+	}
 
-	// TODO: Verify Addr format is [protocol://]host:port
-	addr := d.Get("addr").(string)
+	rootCAsPath, err := getPath(d, "root_ca_path")
+	if err != nil {
+		return nil, diag.FromErr(trace.Wrap(err))
+	}
 
-	tlsConfig, err := loadTLSConfig(certPath, keyPath, rootCAsPath)
+	addr, err := getAddr(d, "addr")
 	if err != nil {
-		// TODO decorate the error into a Terraform-friendly Diags
-		return nil, trace.Wrap(err)
+		return nil, diag.FromErr(trace.Wrap(err))
 	}
 
+	creds := client.LoadKeyPair(certPath, keyPath, rootCAsPath)
+
 	config := client.Config{
 		Addrs:           []string{addr},
 		Dialer:          nil,
 		DialTimeout:     0,
 		KeepAlivePeriod: 0,
 		KeepAliveCount:  0,
-		TLS:             tlsConfig,
+		Credentials:     []client.Credentials{creds},
 	}
 
-	client, err := client.NewClient(config)
+	client, err := client.New(ctx, config)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return nil, diag.FromErr(trace.Wrap(err))
+	}
+
+	return client, diag.Diagnostics{}
+}
+
+// getPath reads path with specified key from provider configuration and checks if it exists
+func getPath(d *schema.ResourceData, key string) (string, error) {
+	p, ok := d.GetOk(key)
+	if !ok {
+		return "", trace.BadParameter("set %v in Teleport porvider configuration", key)
 	}
 
-	return client, nil
+	path, ok := p.(string)
+	if !ok {
+		return "", trace.BadParameter("source Terraform %v key is not a string")
+	}
+
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return path, trace.BadParameter("teleport %v %v file not found", key, path)
+	}
+
+	return path, nil
+}
+
+// getAddr reads API address from config, validates and returns it
+func getAddr(d *schema.ResourceData, key string) (string, error) {
+	a, ok := d.GetOk(key)
+	if !ok {
+		return "", trace.BadParameter("set %v in Teleport provider configuration", key)
+	}
+
+	addr, ok := a.(string)
+	if !ok {
+		return "", trace.BadParameter("please, provide string value for %v in Teleport provider configuration", key)
```

Comment:
This is a very polite error message :) But I'd try to keep error messages for similar scenarios consistent, e.g. above it just says "source Terraform key is not a string".

Reply:
Okay, made it less polite and uniform :-)